tokio-util/src/sync/reusable_box.rs

```diff
@@ -49,62 +48,28 @@ impl<'a, T> ReusableBoxFuture<'a, T> {
     where
         F: Future<Output = T> + Send + 'a,
     {
-        // SAFETY: The pointer is not dangling.
-        let self_layout = {
-            let dyn_future: &(dyn Future<Output = T> + Send) = unsafe { self.boxed.as_ref() };
-            Layout::for_value(dyn_future)
-        };
-
-        if Layout::new::<F>() == self_layout {
-            // SAFETY: We just checked that the layout of F is correct.
-            unsafe {
-                self.set_same_layout(future);
-            }
-
-            Ok(())
-        } else {
-            Err(future)
+        // If we try to inline the contents of this function, the type checker complains because
+        // the bound `T: 'a` is not satisfied in the call to `pending()`. But by putting it in an
+        // inner function that doesn't have `T` as a generic parameter, we implicitly get the bound
+        // `F::Output: 'a` transitively through `F: 'a`, allowing us to call `pending()`.
+        fn real_try_set<'a, F>(
+            this: &mut ReusableBoxFuture<'a, F::Output>,
+            future: F,
+        ) -> Result<(), F>
+        where
+            F: Future + Send + 'a,
+        {
+            // future::Pending<T> is a ZST so this never allocates.
+            let boxed = mem::replace(&mut this.boxed, Box::pin(Pending(PhantomData)));
+            reuse_pin_box(boxed, future, |boxed| this.boxed = Pin::from(boxed))
```

Comment:
this appears to be the only place where `reuse_pin_box` is ever called...do you think it would make sense to just put that code in _this_ function? that way we avoid having an additional function that gets monomorphized over a big pile of generic types, which might help with binary size a bit...

Reply:
Conceptually I quite like the idea of `reuse_pin_box` as an independent futures-unaware utility function — it makes a lot of sense to me as a layer of abstraction. Would putting `#[inline(always)]` on the function have the same effect?

Reply:
My guess is that it'll be inlined either way.

Reply:
Yeah, I think it'll get inlined either way.

I'm not strongly opposed to keeping the separate function; I just wanted to note that it isn't *currently* used elsewhere. :woman_shrugging: